### PR TITLE
Misc Meta improvements

### PR DESCRIPTION
**Pull request by nstarman**
Various improvements:

1. Got rid of MetaStrategyMeta
2. Simplify ``enable_merge_strategies`` context manager

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/utils/metadata/core.py [hunk 1]
```diff
@@ -120,6 +126,7 @@ def __set__(self, instance, value):
             instance._meta = value
 
 
+@dataclass
```
Comment (mhvk) [inline astropy/utils/metadata/core.py:129]: Is this actually useful here? It creates an `__init__`, a `__repr__` we don't use, and defines equality, which I'm not sure we want. I also replaces two lines with two others, adding indirection. Why not just keep it clean as it is?

File: astropy/utils/metadata/merge.py [hunk 2]
```diff
@@ -115,6 +81,33 @@ class method.  This takes ``left`` and ``right`` side arguments and
 
     # types = [(left_types, right_types), ...]
 
+    def __init_subclass__(cls):
+        members = vars(cls)
+
+        # Wrap ``merge`` classmethod to catch any exception and re-raise as
+        # MergeConflictError.
+        if "merge" in members and isinstance(members["merge"], classmethod):
```
Comment (mhvk) [inline astropy/utils/metadata/merge.py:89]: I'd tend to rewrite this a bit to make use of the fact that `cls.merge` may exist:
```
if isinstance((merge := getattr(cls, "merge", None)), classmethod):
```
and then propagate it below.

File: astropy/utils/metadata/merge.py [hunk 3]
```diff
@@ -115,6 +81,33 @@ class method.  This takes ``left`` and ``right`` side arguments and
...
+            def merge(cls, left, right):
+                try:
+                    return orig_merge(cls, left, right)
+                except Exception as err:
+                    raise MergeConflictError(err)
+
+            cls.merge = classmethod(merge)
+
+        # Register merging class (except for base MergeStrategy class)
+        if "types" in members:
```
Comment (mhvk) [inline astropy/utils/metadata/merge.py:102]: Similarly, here
```
if (types := getattr(cls, "types", None)) is not None:
```

**Review by mhvk — COMMENTED** [review]
I like the removal of the metaclass and the contextmanager, and have some suggestions inline to make it even better. But I see no real reason to make `MetaAttribute` a dataclass; would suggest to remove that commit.

File: astropy/utils/metadata/core.py (hunk 1 above)
Comment (nstarman) [reply astropy/utils/metadata/core.py]: ✅

File: astropy/utils/metadata/merge.py (hunk 2 above)
Comment (nstarman) [reply astropy/utils/metadata/merge.py]: ✅

File: astropy/utils/metadata/merge.py (hunk 3 above)
Comment (nstarman) [reply astropy/utils/metadata/merge.py]: ✅

File: astropy/utils/metadata/merge.py (hunk 2 above)
Comment (nstarman) [reply astropy/utils/metadata/merge.py]: That doesn't appear to work.

Comment (nstarman) [reply astropy/utils/metadata/merge.py]: it does when using `members.get()`

**Review by mhvk — APPROVED** [review]
Looks all OK now, thanks!

**Pull request merged**